MsCorePkg/MacAddressEmulationDxe/MacAddressEmulationDxe.c

```diff
@@ -0,0 +1,278 @@
+/** @file
+
+  DXE Driver for handling MAC Address Emulation.
+
+  Copyright (C) Microsoft Corporation.
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+#include "MacAddressEmulationDxe.h"
+
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/DebugLib.h>
+#include <Library/UefiLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/NetLib.h>
+
+#include <Library/MacAddressEmulationPlatformLib.h>
+
+/**
+  @brief Performs sanity checks to ensure an snp can support mac emulation, and ensures that multiple interfaces are not programmed.
+  @param[in] SnpHandle - A handle to an Snp
+  @param[in] Snp - An snp protocol associated with the handle provided
+  @param[in] SnpContext - The snp context created by this driver's entry point
+  @retval TRUE - the SNP supports mac emulation and can be programmed with the emulated address
+  @retval FALSE - the SNP should not be programmed with the emulated address
+
+  @remark  typically called at TPL_NOTIFY
+**/
+BOOLEAN
+SnpSupportsMacEmuCheck (
+  IN CONST EFI_HANDLE                        SnpHandle,
+  IN CONST EFI_SIMPLE_NETWORK_PROTOCOL       *Snp,
+  IN CONST MAC_EMULATION_SNP_NOTIFY_CONTEXT  *Context
+  )
+{
+  BOOLEAN  IsMatch = FALSE;

```

Comment:
I mentioned the goto because it sets the infrastructure in place for clean up logic with roughly the same lines of code now. But, multiple returns would be perfectly fine as well.